opendevin/controller/agent_controller.py

```diff
@@ -196,25 +217,46 @@ async def _step(self):
             await asyncio.sleep(1)
             return
 
-        logger.info(f'STEP {self.state.iteration}', extra={'msg_type': 'STEP'})
-        if self.state.iteration >= self.max_iterations:
-            await self.report_error('Agent reached maximum number of iterations')
-            await self.set_agent_state_to(AgentState.ERROR)
-            return
-
         if self.delegate is not None:
-            delegate_done = await self.delegate._step()
+            await self.delegate._step()
+            delegate_state = self.delegate.get_agent_state()
+            if delegate_state == AgentState.ERROR:
+                # close the delegate upon error
+                await self.delegate.close()
+                await self.report_error('Delegator agent encounters an error')
+                # propagate error state until an agent or user can handle it
+                await self.set_agent_state_to(AgentState.ERROR)
+                return
+            delegate_done = delegate_state == AgentState.FINISHED
             if delegate_done:
+                # retrieve delegate result
                 outputs = self.delegate.state.outputs if self.delegate.state else {}
+
+                # update current controller's state
+                self.state.iteration = self.delegate.state.iteration
+                self.state.num_of_chars = self.delegate.state.num_of_chars
```

Comment:
I'm currently playing with agent delegation, and I think a good design pattern here is that the sub-agent should have no context about the higher-level agent or activity.

For example, I think the sub-agent should start on `iteration=0`. That way we can set `max_iterations` for the delegated task, independent of the higher-level task.

Reply:
> For example, I think the sub-agent should start on iteration=0. That way we can set max_iterations for the delegated task, independent of the higher-level task.

The problem is OpenDevin might never finish because an agent can always delegate again to another agent...